File: src/main/java/toast/client/gui/clickgui/ClickGuiSettings.java
package toast.client.gui.clickgui;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import net.minecraft.client.MinecraftClient;
import toast.client.modules.Module;

import java.awt.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;

import static toast.client.ToastClient.FILE_MANAGER;
import static toast.client.ToastClient.clickGui;

public class ClickGuiSettings {
    public static final String clickguiColorsFile = "guicolors.json";
    public static final String clickguiPosFile = "clickGuiPositions.json";
    public static final int defaultOnTextColor = new Color(255, 255, 255, 255).getRGB();
    public static final int defaultOffTextColor = new Color(177, 177, 177, 255).getRGB();
    public static final int defaultBgColor = new Color(0, 0, 0, 64).getRGB();
    public static final int defaultHoverBgColor = new Color(131, 212, 252, 92).getRGB();
    public static final int defaultBoxColor = new Color(0, 0, 0, 255).getRGB();
    public static final int defaultPrefixColor = new Color(8, 189, 8, 255).getRGB();
    public static final int defaultClickColor = new Color(121, 205, 255, 128).getRGB();
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    public static Map<String, CategorySetting> categoryPositions = new TreeMap<>();
    public Colors colors = new Colors();

    public CategorySetting getPositions(String category) {
        if (categoryPositions == null) {
            loadPositions();
        }
        return categoryPositions.get(category);
    }

    public void initCategoryPositions() {
        categoryPositions = new TreeMap<>();
        int i = 0;
        int y = 5;
        System.out.println(MinecraftClient.getInstance().getWindow().getWidth());
        for (Module.Category category : Module.Category.values()) {
            int x = 5 + (ClickGuiScreen.width * i) + (5 * i);
            if (x + ClickGuiScreen.width + 10 > MinecraftClient.getInstance().getWindow().getWidth() / 2) {
                y = y + (MinecraftClient.getInstance().getWindow().getHeight() / 2) / 3;
                x = 5;
                i = 0;
            }
            categoryPositions.put(category.toString(), new CategorySetting(x, y, false, new ArrayList<>()));
            i++;
        }
    }

    public Colors getColors() {
        return colors;
    }

    public void loadPositions() {
        try {
            categoryPositions = gson.fromJson(new FileReader(FILE_MANAGER.createFile(new File(clickguiPosFile))), new TypeToken<Map<String, CategorySetting>>() {
            }.getType());
            if (categoryPositions == null && clickGui != null) {
                initCategoryPositions();
                savePositions();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public void savePositions() {
        FILE_MANAGER.writeFile(new File(clickguiPosFile), gson.toJson(categoryPositions));
    }

    public void saveColors() {
        FILE_MANAGER.writeFile(new File(clickguiColorsFile), gson.toJson(colors));
    }

    public void allDefaults() {
        colors = new Colors();
        setCategoryDefaults();
        setDescriptionDefaults();
        setModuleDefaults();
        setSettingsDefaults();
    }

    public void setCategoryDefaults() {
        colors.categoryBgColor = defaultBgColor;
        colors.categoryBoxColor = defaultBoxColor;
        colors.categoryClickColor = defaultClickColor;
        colors.categoryHoverBgColor = defaultHoverBgColor;
        colors.categoryPrefix = "> ";
        colors.categoryPrefixColor = defaultPrefixColor;
        colors.categoryTextColor = defaultOnTextColor;
    }

    public void setDescriptionDefaults() {
        colors.descriptionBgColor = new Color(83, 83, 83, 255).getRGB();
        colors.descriptionBoxColor = defaultBoxColor;
        colors.descriptionPrefix = "";
        colors.descriptionPrefixColor = defaultPrefixColor;
        colors.descriptionTextColor = defaultOnTextColor;
    }

    public void setModuleDefaults() {
        colors.moduleBoxColor = defaultBoxColor;
        colors.moduleClickColor = defaultClickColor;
        colors.moduleHoverBgColor = defaultHoverBgColor;
        colors.moduleOffBgColor = defaultBgColor;
        colors.moduleOffTextColor = defaultOffTextColor;
        colors.moduleOnBgColor = defaultBgColor;
        colors.moduleOnTextColor = defaultOnTextColor;
        colors.modulePrefix = " > ";
        colors.modulePrefixColor = defaultPrefixColor;
    }

    public void setSettingsDefaults() {
        colors.settingBoxColor = defaultBoxColor;
        colors.settingClickColor = defaultClickColor;
        colors.settingHoverBgColor = defaultHoverBgColor;
        colors.settingOffBgColor = defaultBgColor;
        colors.settingOffTextColor = defaultOffTextColor;
        colors.settingOnBgColor = defaultBgColor;
        colors.settingOnTextColor = defaultOnTextColor;
        colors.settingSliderBarColor = defaultOffTextColor;
        colors.settingSliderKnobColor = defaultOnTextColor;
        colors.settingSliderKnobHoverColor = defaultOnTextColor;
        colors.settingSliderSideNumbersColor = defaultOnTextColor;
        colors.settingPrefix = "    ";
        colors.settingPrefixColor = defaultPrefixColor;
    }

    public void loadColors() {
        try {
            colors = gson.fromJson(new FileReader(FILE_MANAGER.createFile(new File(clickguiColorsFile))), new TypeToken<Colors>() {
            }.getType());
            if (colors == null) {
                allDefaults();
                saveColors();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    public static class Colors {
        public String categoryPrefix;
        public int categoryTextColor;
        public int categoryHoverBgColor;
        public int categoryBgColor;
        public int categoryBoxColor;
        public int categoryPrefixColor;
        public int categoryClickColor;
        public String descriptionPrefix;
        public int descriptionTextColor;
        public int descriptionBoxColor;
        public int descriptionBgColor;
        public int descriptionPrefixColor;
        public String modulePrefix;
        public int moduleOnTextColor;
        public int moduleOnBgColor;
        public int moduleOffTextColor;
        public int moduleOffBgColor;
        public int moduleHoverBgColor;
        public int moduleBoxColor;
        public int modulePrefixColor;
        public int moduleClickColor;
        public String settingPrefix;
        public int settingOnTextColor;
        public int settingSliderBarColor;
        public int settingSliderKnobColor;
        public int settingSliderKnobHoverColor;
        public int settingSliderSideNumbersColor;
        public int settingOnBgColor;
        public int settingOffTextColor;
        public int settingOffBgColor;
        public int settingHoverBgColor;
        public int settingBoxColor;
        public int settingPrefixColor;
        public int settingClickColor;
    }
}

File: src/main/java/dev/toastmc/client/gui/click/ClickGuiSettings.kt
package dev.toastmc.client.gui.click

import com.google.gson.reflect.TypeToken
import dev.toastmc.client.module.Category
import dev.toastmc.client.module.Module
import dev.toastmc.client.util.gson
import dev.toastmc.client.util.mc
import net.minecraft.client.MinecraftClient
import java.awt.Color
import java.io.File
import java.io.FileNotFoundException
import java.util.*

open class ClickGuiSettings(private val clickGuiScreen: ClickGuiScreen) {
    private val clickColorsFile = File("toastclient/colors.json")
    private val clickPosFile = File("toastclient/clickgui.json")
    private val defaultOnTextColor = Color(255, 255, 255, 255).rgb
    private val defaultOffTextColor = Color(177, 177, 177, 255).rgb
    private val defaultOffBgColor = Color(0, 0, 0, 64).rgb
    private val defaultOnBgColor = Color(0, 0, 0, 155).rgb
    private val defaultHoverBgColor = Color(131, 212, 252, 92).rgb
    private val defaultBoxColor = Color(0, 0, 0, 255).rgb
    private val defaultPrefixColor = Color(255, 0, 0, 255).rgb
    private val defaultClickColor = Color(121, 205, 255, 128).rgb
    private var categoryPositions: MutableMap<String, CategorySetting> = TreeMap()
    var colors = Colors()

    init {
        loadColors()
        loadPositions()
    }

    fun getPositions(category: String): CategorySetting {
        var setting = categoryPositions[category]
        if (setting == null) {
            loadPositions()
            setting = categoryPositions[category]
            if (setting == null) initCategoryPositions()
        }
        return categoryPositions[category]!!
    }

    fun initCategoryPositions() {
        if (mc.window == null) return
        var i = 0
        var y = 5
        for (category in Category.values()) {
            if (category == Category.NONE) continue
            var x: Int = 5 + clickGuiScreen.w.times(i) + 5 * i
            if (x + clickGuiScreen.w + 30 > MinecraftClient.getInstance().window.width / 2) {
                y += MinecraftClient.getInstance().window.height / 2 / 3
                x = 100
                i = 0
            }
            (categoryPositions as TreeMap<String, CategorySetting>)[category.toString()] = CategorySetting(x.toDouble(), y.toDouble(), false, ArrayList())
            i++
        }
    }

    fun loadPositions() {
        try {
            categoryPositions = TreeMap()
            val loadedData: MutableMap<String?, CategorySetting?>? = gson.fromJson(clickPosFile.readText(), object : TypeToken<Map<String?, CategorySetting?>?>() {}.type)
            if (loadedData != null) {
                for (item in loadedData) if (item.key != null || item.value != null) categoryPositions.putIfAbsent(
                    item.key!!,
                    item.value!!
                )
                return
            }
        } catch (_: FileNotFoundException) {
            println("Failed to load GUI positions")
        } finally {
            initCategoryPositions()
            savePositions()
        }
    }

    fun moduleExpanded(cat: String, mod: Module): Boolean {
        return getPositions(cat).expandedModules.contains(mod.label)
    }


    open fun savePositions() {
        clickPosFile.createNewFile()
        clickPosFile.writeText(gson.toJson(categoryPositions))
    }

    fun saveColors() {
        clickColorsFile.createNewFile()
        clickColorsFile.writeText(gson.toJson(colors))
    }

    private fun allDefaults() {
        colors = Colors()
        setCategoryDefaults()
        setDescriptionDefaults()
        setModuleDefaults()
        setSettingsDefaults()
    }

    private fun setCategoryDefaults() {
        colors.categoryBgColor = defaultOffBgColor
        colors.categoryBoxColor = defaultBoxColor
        colors.categoryClickColor = defaultClickColor
        colors.categoryHoverBgColor = defaultHoverBgColor
        colors.categoryPrefix = "+ "
        colors.categoryPrefixColor = defaultPrefixColor
        colors.categoryTextColor = defaultOnTextColor
    }

    private fun setDescriptionDefaults() {
        colors.descriptionBgColor = Color(83, 83, 83, 255).rgb
        colors.descriptionBoxColor = defaultBoxColor
        colors.descriptionPrefix = ""
        colors.descriptionPrefixColor = defaultPrefixColor
        colors.descriptionTextColor = defaultOnTextColor
    }

    private fun setModuleDefaults() {
        colors.moduleBoxColor = defaultBoxColor
        colors.moduleClickColor = defaultClickColor
        colors.moduleHoverBgColor = defaultHoverBgColor
        colors.moduleOffBgColor = defaultOffBgColor
        colors.moduleOffTextColor = defaultOffTextColor
        colors.moduleOnBgColor = defaultOnBgColor
        colors.moduleExpandadBgColor = defaultOnBgColor
        colors.moduleOnTextColor = defaultOnTextColor
        colors.modulePrefix = " > "
        colors.modulePrefixColor = defaultPrefixColor
    }

    private fun setSettingsDefaults() {
        colors.settingBoxColor = defaultBoxColor
        colors.settingClickColor = defaultClickColor
        colors.settingHoverBgColor = defaultHoverBgColor
        colors.settingOffBgColor = defaultOnBgColor
        colors.settingOffTextColor = defaultOffTextColor
        colors.settingOnBgColor = defaultOnBgColor
        colors.settingOnTextColor = defaultOnTextColor
        colors.settingSliderBarColor = defaultOffTextColor
        colors.settingSliderKnobColor = defaultOnTextColor
        colors.settingSliderKnobHoverColor = defaultOnTextColor
        colors.settingSliderSideNumbersColor = defaultOnTextColor
        colors.settingPrefix = " "
        colors.settingPrefixColor = defaultPrefixColor
    }

    fun loadColors() {
        try {
            val newColors: Colors? = gson.fromJson(clickColorsFile.readText(), object : TypeToken<Colors?>() {}.type)
            if (newColors != null) {
                colors = newColors
                return
            }
        } catch (_: FileNotFoundException) {
        } finally {
            allDefaults()
            saveColors()
        }
    }

    class Colors {
        var categoryPrefix = ""
        var categoryTextColor = 0
        var categoryHoverBgColor = 0
        var categoryBgColor = 0
        var categoryBoxColor = 0
        var categoryPrefixColor = 0
        var categoryClickColor = 0
        var descriptionPrefix = ""
        var descriptionTextColor = 0
        var descriptionBoxColor = 0
        var descriptionBgColor = 0
        var descriptionPrefixColor = 0
        var modulePrefix = ""
        var moduleOnTextColor = 0
        var moduleOnBgColor = 0
        var moduleExpandadBgColor = 0
        var moduleOffTextColor = 0
        var moduleOffBgColor = 0
        var moduleHoverBgColor = 0
        var moduleBoxColor = 0
        var modulePrefixColor = 0
        var moduleClickColor = 0
        var settingPrefix = ""
        var settingOnTextColor = 0
        var settingSliderBarColor = 0
        var settingSliderKnobColor = 0
        var settingSliderKnobHoverColor = 0
        var settingSliderSideNumbersColor = 0
        var settingOnBgColor = 0
        var settingOffTextColor = 0
        var settingOffBgColor = 0
        var settingHoverBgColor = 0
        var settingBoxColor = 0
        var settingPrefixColor = 0
        var settingClickColor = 0
    }
}